src/Interpreters/TransactionVersionMetadata.cpp

```diff
@@ -0,0 +1,376 @@
+    /// It means that some transaction is creating/removing the part right now or has done it recently
+    /// and we don't know if it was already committed or not.
+    assert(!had_mincsn || (had_maxtid && !had_maxcsn));
+    assert(current_tid.isEmpty() || (mintid != current_tid && max_lock != current_tid.getHash()));
+
+    /// Before doing CSN lookup, let's check some extra conditions.
+    /// If snapshot_version <= some_tid.start_csn, then changes of the transaction with some_tid
+    /// are definitely not visible for us (because the transaction can be committed with greater CSN only),
+    /// so we don't need to check if it was committed.
+    if (snapshot_version <= mintid.start_csn)
+        return false;
+
+    /// Check if mintid/maxtid transactions are committed and write CSNs
+    /// TODO Transactions: we probably need some optimizations here
+    /// to avoid some CSN lookups or make the lookups cheaper.
+    /// NOTE: Old enough committed parts always have written CSNs,
+    /// so we can determine their visibility through fast path.
+    /// But for long-running writing transactions we will always do
+    /// CNS lookup and get 0 (UnknownCSN) until the transaction is committer/rolled back.
+    min = TransactionLog::instance().getCSN(mintid);
+    if (!min)
+        return false;   /// Part creation is not committed yet
+
+    /// We don't need to check if CSNs are already written or not,
+    /// because once written CSN cannot be changed, so it's safe to overwrite it (with the same value).
+    mincsn.store(min, std::memory_order_relaxed);
+
+    if (max_lock)
+    {
+        max = TransactionLog::instance().getCSN(max_lock);
+        if (max)
+            maxcsn.store(max, std::memory_order_relaxed);
+    }
+
+    return min <= snapshot_version && (!max || snapshot_version < max);
+}
+
+bool VersionMetadata::canBeRemoved(Snapshot oldest_snapshot_version)
+{
+    CSN min = mincsn.load(std::memory_order_relaxed);
+    /// We can safely remove part if its creation was rolled back
+    if (min == Tx::RolledBackCSN)
+        return true;
+
+    if (!min)
+    {
+        /// Cannot remove part if its creation not committed yet
+        min = TransactionLog::instance().getCSN(mintid);
+        if (min)
+            mincsn.store(min, std::memory_order_relaxed);
+        else
+            return false;
+    }
+
+    /// Part is probably visible for some transactions (part is too new or the oldest snapshot is too old)
+    if (oldest_snapshot_version < min)
+        return false;
+
+    TIDHash max_lock = maxtid_lock.load(std::memory_order_relaxed);
+    /// Part is active
+    if (!max_lock)
+        return false;
+
+    CSN max = maxcsn.load(std::memory_order_relaxed);
+    if (!max)
+    {
+        /// Part removal is not committed yet
+        max = TransactionLog::instance().getCSN(max_lock);
+        if (max)
+            maxcsn.store(max, std::memory_order_relaxed);
```

Comment:
There's a short period of time when transaction is committed, but `csn` is not set yet. We can set it here to avoid extra CSN lookups in concurrent SELECTs.